libwfssl/src/ssl.c

```diff
@@ -534,7 +547,33 @@ WF_OPENSSL(jint, initialize) (JNIEnv *e, jobject o, jstring libCryptoPath, jstri
 
     ssl_thread_setup();
 
-    /* TODO: engine support? */
+    if (customEngine != NULL) {
+        if (strcmp(engine, "auto") == 0) {
+            crypto_methods.ENGINE_register_all_complete();
+        } else {
+            ENGINE *ssl_engine = crypto_methods.ENGINE_by_id(engine);
+            if (ssl_engine == NULL) {
+                ssl_engine = crypto_methods.ENGINE_by_id("dynamic");
+                if (ssl_engine) {
+                    if (!crypto_methods.ENGINE_ctrl_cmd_string(ssl_engine, "SO_PATH", engine, 0) || !crypto_methods.ENGINE_ctrl_cmd_string(ssl_engine, "LOAD", NULL, 0)) {
+                        crypto_methods.ENGINE_free(ssl_engine);
+                        ssl_engine = NULL;
+                        tcn_Throw(e, "Could not load openssl custom engine (%s) .so library file", engine);
+                        goto init_failed;
+                    }
+                }
+            }
+
+            if (ssl_engine) {
```

Comment:
For this case, it looks like both the Netty and Tomcat OpenSSL implementations also do this before attempting to set the default:

```
#ifdef ENGINE_CTRL_CHIL_SET_FORKCHECK
                if (strcmp(J2S(engine), "chil") == 0)
                    ENGINE_ctrl(tcn_ssl_engine, ENGINE_CTRL_CHIL_SET_FORKCHECK, 1, 0, 0);
#endif
```

See https://github.com/netty/netty-tcnative/blob/main/openssl-dynamic/src/main/c/ssl.c#L804-L810 and https://github.com/apache/tomcat-native/blob/main/native/src/ssl.c#L540-L547.

Do you think we could do something similar here?

Reply:
I don't know the real reason to add this logic. So I need to check.

Reply:
This logic is specifically for the "chil" engine. It's not strictly required to add this for your changes but since both the Netty and Tomcat OpenSSL implementations have this, I think we might as well add the same logic here as well.

Reply:
The `ENGINE_CTRL_CHIL_SET_FORKCHECK` is never going to be defined, so adding the lines like they are now is doing nothing. Note that wildfly-openssl works using `dlopen` without any openssl include at compilation. If we really want to set this option for `chil` like in the other implementations we need to add the define in `wfssl.h`:

```c
+/* Flags specific to the nCipher "chil" engine */
+# define ENGINE_CTRL_CHIL_SET_FORKCHECK          100
```

And always execute the `ENGINE_ctrl` in case of the `chil` engine.

Nit-picky request, please add the brackets to the if. :smile:

Reply:
@heyuanliu-intel The `#ifdef` is not needed now, it's going to be always defined. All the rest LGTM now. Thanks!

Reply:
okay. I will remove it.